caravel/views.py

```diff
@@ -971,15 +992,113 @@ def msg(self):
+
+        duplicates = (
+            db.session.query(models.DatasourceAccessRequest)
+            .filter(
+                models.DatasourceAccessRequest.created_by_fk ==
+                g.user.get_id(),
+                models.DatasourceAccessRequest.table_id == table_id,
+                models.DatasourceAccessRequest.druid_datasource_id ==
+                druid_datasource_id,
+            ).all()
+        )
+
+        if duplicates:
+            flash(__(
+                "You have already requested access to the datasource %(name)s",
+                name=datasource_name), "warning")
+            return redirect('/slicemodelview/list/')
+
+        access_request = models.DatasourceAccessRequest(
+            table_id=table_id,
+            druid_datasource_id=druid_datasource_id,
+        )
+        db.session.add(access_request)
+        db.session.commit()
+        flash(__("Access to the datasource %(name)s was requested",
+                 name=datasource_name), "info")
+        return redirect('/slicemodelview/list/')
+
+    @log_this
+    @has_access
+    @expose("/approve")
+    def approve(self):
+        request_access_id = request.args.get('request_access_id')
+        role_to_grant = request.args.get('role_to_grant')
+        role_to_extend = request.args.get('role_to_extend')
+        access_request = (
+            db.session.query(models.DatasourceAccessRequest)
+            .filter_by(id=request_access_id).first()
+        )
+        # check if you can approve
+        if (self.all_datasource_access() or
+                (access_request.datasource and
+                 access_request.datasource.created_by_fk and
+                 g.user.id == access_request.datasource.owner_id)):
+            # can by done by admin only
+            if role_to_grant:
+                role = sm.find_role(role_to_grant)
+                access_request.created_by.roles.append(role)
+                flash(__(
+                    "%(user)s was granted the role %(role)s that gives access "
+                    "to the %(datasource)s",
+                    user=access_request.creator(),
+                    role=role_to_grant,
+                    datasource=access_request.datasource.full_name), "info")
+
+            if role_to_extend:
+                sm.add_permission_role(sm.find_role(role_to_extend),
+                                       access_request.permission_view)
+                flash(__("Role %(role)s was extended to provide the access to"
+                         " the datasource %(datasource)s",
+                         role=role_to_extend,
+                         datasource=access_request.datasource.full_name),
+                      "info")
+
+        else:
+            flash(__("You have no permission to approve this request"),
+                  "danger")
+            return redirect('/accessrequestsmodelview/list/')
+
+        db.session.delete(access_request)
```

Comment:
I should have caught this before, but it's critical that we keep an audit trail of who granted access to what/who/when. One way to do this is to not delete the request, and flag it as resolved instead (add a state field to the model, a fk to the userid who granted access, timestamp when it was granted, and maybe the resolution that was applied). 

Another option is to store the operation in another model.